_posts/2022-03-21-getting-started-with-protobufs-for-typescript.md

```diff
@@ -0,0 +1,254 @@
+        },
+      ],
+    },
+    {
+      name: "Jane Smith",
+      phones: [
+        {
+          phoneNumber: "0987654321",
+          phoneType: Person_PhoneType.HOME,
+        },
+      ],
+    },
+  ],
+};
+```
+This is almost identical to the original JSON object we used in the [data structure section above](#define-the-data-structure-for-your-messages), how fantastic is that!  There are two differences in the above that take advantage of some awesome idiomatic Typescript and make life really easy:
+- The object type assignment `AddressBook` which allows you to use type checking for creating valid messages
+- The `phoneType` is assigned through an enum, e.g. `Person_PhoneType.MOBILE`
+
+However, if at this point you are thinking _*"Now wait up, this is not like the other tutorials I have seen. I was expecting to use a class object and function calls to assign the data fields"*_, you can do it that way with Typescript, have a look at [How do do this so it looks like most published tutorials](#how-do-do-this-so-it-looks-like-most-published-tutorials) section below.
+
+Congrats, this is the hard work done. You are most of the way there!
+
+## Serialise the Message to Protobuf
+
+After all of the above, this is the easy bit. Again taking instruction from the generated Types file:
+
+- `const serializedMessage = AddressBook.encode(myAddressBook).finish()`
+
+## Do Something with the Serialised Message
+
+The how of transmitting the encoded messages is out of scope here, but you can for example transmit over gRPC or HTTP or even save to file if you really want (as a lot of tutorials unhelpfully seem to do).
+
+## De-serialise the Message
+
+After you have transmitted or stored your message, at some point you are going to want to use its contents, so you are going to have to de-serialize. Again this is really easy:
+
+- `const deSerializedMessage = AddressBook.decode(proto)`
+
+Done! You have your original `AddressBook` Typed message object again.
+
+Congrats, you survived and don’t need to spend days bashing your head against the wall.
+
+
+# Closing Thoughts
+
+Getting to even this starting point of being able to do the basics with Protobufs, as above, was not a straightforward journey for me and I hope if you have read this far it might shortcut that process for you.  I am sure the docs will get better over time but Protobufs are not a beginner topic and everyone has to start somewhere.
+
+You can now:
+1) Define you own message structure
+2) Generate TypeScript `.ts` output for working with Protobufs within your code
+2) Serialize messages into Protobuf for transmitting
+3) De-serialise Protobufs back in to useable message objects.
+
+For a next step, I would recommend spending spending some time understanding the [proto3 data structure](https://developers.google.com/protocol-buffers/docs/proto3) so that you can more easily build you own message structures.
+
+
+# How do do this so it looks like most published tutorials
```

Comment:
# How do do this so it looks like most published tutorials
should probably be
# How to do this so it looks like most published tutorials